File: app/src/main/java/com/github/gotify/messages/provider/MessageFacade.java
package com.github.gotify.messages.provider;

import com.github.gotify.client.api.MessageApi;
import com.github.gotify.client.model.Message;
import com.github.gotify.client.model.PagedMessages;
import java.util.List;

public class MessageFacade {
    private final ApplicationHolder applicationHolder;
    private final MessageRequester requester;
    private final MessageStateHolder state;
    private final MessageImageCombiner combiner;

    public MessageFacade(MessageApi api, ApplicationHolder applicationHolder) {
        this.applicationHolder = applicationHolder;
        this.requester = new MessageRequester(api);
        this.combiner = new MessageImageCombiner();
        this.state = new MessageStateHolder();
    }

    public synchronized List<MessageWithImage> get(Integer appId) {
        return combiner.combine(state.state(appId).messages, applicationHolder.get());
    }

    public synchronized void addMessages(List<Message> messages) {
        for (Message message : messages) {
            state.newMessage(message);
        }
    }

    public synchronized List<MessageWithImage> loadMore(Integer appId) {
        MessageState state = this.state.state(appId);
        if (state.hasNext || !state.loaded) {
            PagedMessages pagedMessages = requester.loadMore(state);
            this.state.newMessages(appId, pagedMessages);
        }
        return get(appId);
    }

    public synchronized void loadMoreIfNotPresent(Integer appId) {
        MessageState state = this.state.state(appId);
        if (!state.loaded) {
            loadMore(appId);
        }
    }

    public synchronized void clear() {
        this.state.clear();
    }

    public int getLastReceivedMessage() {
        return state.getLastReceivedMessage();
    }

    public synchronized void deleteLocal(Message message) {
        // If there is already a deletion pending, that one should be executed before scheduling the
        // next deletion.
        if (this.state.deletionPending()) commitDelete();
        this.state.deleteMessage(message);
    }

    public synchronized void commitDelete() {
        if (this.state.deletionPending()) {
            MessageDeletion deletion = this.state.purgePendingDeletion();
            this.requester.asyncRemoveMessage(deletion.getMessage());
        }
    }

    public synchronized MessageDeletion undoDeleteLocal() {
        return this.state.undoPendingDeletion();
    }

    public synchronized boolean deleteAll(Integer appId) {
        boolean success = this.requester.deleteAll(appId);
        this.state.deleteAll(appId);
        return success;
    }

    public synchronized boolean canLoadMore(Integer appId) {
        return state.state(appId).hasNext;
    }
}

File: app/src/main/kotlin/com/github/gotify/messages/provider/MessageFacade.kt
package com.github.gotify.messages.provider

import com.github.gotify.client.api.MessageApi
import com.github.gotify.client.model.Message

internal class MessageFacade(api: MessageApi, private val applicationHolder: ApplicationHolder) {
    private val requester = MessageRequester(api)
    private val state = MessageStateHolder()

    @Synchronized
    operator fun get(appId: Long): List<MessageWithImage> {
        return MessageImageCombiner.combine(state.state(appId).messages, applicationHolder.get())
    }

    @Synchronized
    fun addMessages(messages: List<Message>) {
        messages.forEach {
            state.newMessage(it)
        }
    }

    @Synchronized
    fun loadMore(appId: Long): List<MessageWithImage> {
        val state = state.state(appId)
        if (state.hasNext || !state.loaded) {
            val pagedMessages = requester.loadMore(state)
            if (pagedMessages != null) {
                this.state.newMessages(appId, pagedMessages)
            }
        }
        return get(appId)
    }

    @Synchronized
    fun loadMoreIfNotPresent(appId: Long) {
        val state = state.state(appId)
        if (!state.loaded) {
            loadMore(appId)
        }
    }

    @Synchronized
    fun clear() {
        state.clear()
    }

    fun getLastReceivedMessage(): Long = state.lastReceivedMessage

    @Synchronized
    fun deleteLocal(message: Message) {
        // If there is already a deletion pending, that one should be executed before scheduling the
        // next deletion.
        if (state.deletionPending()) commitDelete()
        state.deleteMessage(message)
    }

    @Synchronized
    fun commitDelete() {
        if (state.deletionPending()) {
            val deletion = state.purgePendingDeletion()
            requester.asyncRemoveMessage(deletion!!.message)
        }
    }

    @Synchronized
    fun undoDeleteLocal(): MessageDeletion? = state.undoPendingDeletion()

    @Synchronized
    fun deleteAll(appId: Long): Boolean {
        val success = requester.deleteAll(appId)
        state.deleteAll(appId)
        return success
    }

    @Synchronized
    fun canLoadMore(appId: Long): Boolean = state.state(appId).hasNext
}
